File: c++/004-MedianOfTwoSortedArrays.cpp
#include<iostream>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size(), m = nums2.size();
        if ((n + m) & 1) {
            return findMedian(nums1.begin(), nums2.begin(), n, m, (n + m + 1) >> 1);
        } else {
            return double(findMedian(nums1.begin(), nums2.begin(), n, m, (n + m) >> 1) + 
                findMedian(nums1.begin(), nums2.begin(), n, m, (n + m + 2) >> 1)) / 2.0;
        }
    }
    
    int findMedian(vector<int>::iterator num1, vector<int>::iterator num2, int n, int m, int k) {
        if (n > m) return findMedian(num2, num1, m, n, k);
        else
        if (n == 0) return *(num2 + k - 1);
        else
        if (k == 1) return min(*num1, *num2);
        else {
            int n1 = (k / 2 <= n) ? k / 2 : n;
            int m1 = k - n1;
            if ((*(num1 + n1 - 1)) < (*(num2 + m1 - 1))) {
                return findMedian(num1 + n1, num2, n - n1, m, k - n1);
            } else if ((*(num1 + n1 - 1)) > (*(num2 + m1 - 1))) {
                return findMedian(num1, num2 + m1, n, m - m1, k - m1);
            } else {
                return *(num1 + n1 - 1);
            }
        }
    }
};

int main() {
	return 0;
}

File: python/004-MedianofTwoSortedArrays.py
class Solution:
    # @param {integer[]} nums1
    # @param {integer[]} nums2
    # @return {float}
    def findMedianSortedArrays(self, nums1, nums2):
        n, m = len(nums1), len(nums2)
        if (n + m) % 2 == 1 :
            return self.findMedian(nums1, nums2, 0, 0, int((n + m) / 2 + 1))
        else :
            return (self.findMedian(nums1, nums2, 0, 0, int((n + m) / 2)) + \
                self.findMedian(nums1, nums2, 0, 0, int((n + m) / 2 + 1))) / 2.0
    
    def findMedian(self, nums1, nums2, st1, st2, k) :
        n1, n2 = len(nums1) - st1, len(nums2) - st2
        if n1 > n2 :
            return self.findMedian(nums2, nums1, st2, st1, k)
        elif n1 == 0 :
            return nums2[st2 + k - 1]
        elif k == 1 :
            return min(nums1[st1], nums2[st2])
        else :
            halfn1 = min(n1, int(k / 2))
            halfn2 = k - halfn1
            if nums1[st1 + halfn1 - 1] <= nums2[st2 + halfn2 - 1] :
                return self.findMedian(nums1, nums2, st1 + halfn1, st2, k - halfn1)
            else :
                return self.findMedian(nums1, nums2, st1, st2 + halfn2, k - halfn2)
            
nums1 = [1, 3, 5]
nums2 = [2, 4, 6]
solution = Solution()
print(solution.findMedianSortedArrays(nums1, nums2))